Create cr command: change and add integration tests

hzhou97 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Changes the structure of the integration tests for the create cr command in order to have more fine grained testing.
Replaces PR [3086](https://github.com/jetstack/cert-manager/pull/3086) because of rebasing issues.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```

/kind cleanup
/area testing

hzhou97 commented:
/assign @wallrj

wallrj commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,239 @@ import (
+				CertFileName:     test.certFilename,
+			}
+
+			// Validating args and flags
+			err := opts.Validate(test.inputArgs)
+			if err != nil {
+				if !test.expValidateErr {
+					t.Errorf("got unexpected error when validating args and flags: %v", err)
+				} else if err.Error() != test.expErrMsg {
+					t.Errorf("got unexpected error when validating args and flags, expected: %v; actual: %v", test.expErrMsg, err)
+				}
+				return
+			} else {
+				// got no error
+				if test.expValidateErr {
+					t.Errorf("expected but got no error validating args and flags")
+				}
+			}
```
I would add a TODO item suggesting we follow this up by adding unit tests for the `opts.Validate` in the package where it is defined.
And generally, I'd try and do more unit tests for error conditions and leave integration tests for mostly success cases.

wallrj commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -288,31 +464,31 @@ func TestCtlCreateCR(t *testing.T) {
 	}
 }
 
+func getTestDataDir(t *testing.T) string {
+	testWorkingDirectory, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	return testWorkingDirectory + "/testdata/"
+}
+
```
I'm not sure if we need this. `go test` will automatically set the working-dir to match the package path under test, so `testdata/` can always be reached in tests using that relative path. See https://dave.cheney.net/2016/05/10/test-fixtures-in-go

hzhou97 replied:
We need this for this test because we cd into a tmp directory for each test for easier clean-up of the generated files.

meyskens replied:
Is this a tmp directory?

hzhou97 replied:
The `testdata` dir is not a tmp directory, but the working directory for each test is a tmp directopry that is set up by `setupPathForTest`

hzhou97 commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,239 @@ import (
+			keyFilename:    "",
+			expValidateErr: false,
+			expRunErr:      true,
+			expErrMsg:      "error when getting infos out of Result: the namespace from the provided object \"testns-1\" does not match the namespace \"testns-2\". You must pass '--namespace=testns-1' to perform this operation.",
```
I removed `error when getting infos out of Result:`, but the rest of the error is returned from another function, so I can't really change it.

hzhou97 commented:
I refactored some parts, and moved all the checking for errors returned by `Validate` to unit testing. I wanted to refactor the whole of `TestCtlCreateCRBehaviourBeforeAnythingIsCreated` to unit tests, but the cases that are still there require a manifest, and I didn't think of a good way to share the files/add a new file for the unit tests.

/assign @wallrj

hzhou97 commented:
/retest

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest.go:
```diff
@@ -192,13 +196,13 @@ func (o *Options) Run(args []string) error {
 		FilenameParam(o.EnforceNamespace, &resource.FilenameOptions{Filenames: []string{o.InputFilename}}).Flatten().Do()
 
 	if err := r.Err(); err != nil {
-		return fmt.Errorf("error when getting Result from Builder: %s", err)
```
Any reason we stopped wrapping these errors?

hzhou97 replied:
Richard mentioned they made it harder to understand the error, and I agreed

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:
```diff
@@ -0,0 +1,102 @@
+		expErrMsg string
+	}{
+		"CR name not passed as arg throws error": {
+			inputFile: "example.yaml",
```
As discussed in the standup, would be maybe better not to read these from disk.

meyskens replied:
Sorry, misread this is the validation function

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:
```diff
@@ -0,0 +1,102 @@
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			opts := &Options{
```
How about moving options as a parameter of the table test itself instead of making this object every time

hzhou97 replied:
I don't see how what the benefit would be? And this way it's consistent with the integration test.

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:
```diff
@@ -0,0 +1,102 @@
+				} else if err.Error() != test.expErrMsg {
+					t.Errorf("got unexpected error when validating args and flags, expected: %v; actual: %v", test.expErrMsg, err)
+				}
+				return
```
You can use a t.Fatal here instead

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:
```diff
@@ -0,0 +1,102 @@
+
+			// Validating args and flags
+			err := opts.Validate(test.inputArgs)
+			if err != nil {
```
Not a too big fan of nested `if` here

hzhou97 replied:
I'm not sure what a better option would be?

meyskens commented on cmd/ctl/pkg/create/certificaterequest/certificaterequest_test.go:
```diff
@@ -0,0 +1,102 @@
+				return
+			} else {
+				// got no error
+				if test.expErr {
```
`else if` could work better here

meyskens commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,203 @@ import (
+			defer cleanUpFunc()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
```
Same comment as before?

hzhou97 replied:
Oh right, you wanted to change both the unit test and the integration tests. Same as before, I don't see how one way is better than the other.

meyskens commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,203 @@ import (
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			cleanUpFunc := setupPathForTest(t)
```
I think we can make this one line `setupPathForTest(t)()`

hzhou97 replied:
I think it's more readable the way it is?

meyskens commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,203 @@ import (
+
+			// Create CR
+			err = opts.Run(test.inputArgs)
+			if err != nil {
```
Same as before with nested ifs

meyskens commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -35,7 +36,203 @@ import (
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			cleanUpFunc := setupPathForTest(t)
```
Same as before

meyskens commented on test/integration/ctl/ctl_create_cr_test.go:
```diff
@@ -288,31 +422,31 @@ func TestCtlCreateCR(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	return testWorkingDirectory + "/testdata/"
```
we have `path.Join()` in Go which does this for you with working around weird edge causes (cough windows)

meyskens commented:
Following @wallrj 

/lgtm
/approve
